File: javascript/leetcode/1047. Remove All Adjacent Duplicates In String.js
/**
 * @param {string} S
 * @return {string}
 */
var removeDuplicates = function (S) {
  const seen = [];
  seen.push(S[0]);

  for (let i = 1; i < S.length; ++i) {
    if (seen.length && seen[seen.length - 1] === S[i]) seen.pop();
    else seen.push(S[i]);
  }

  return seen.join("");
};

File: c++/leetcode/1047. Remove All Adjacent Duplicates In String.cpp
class Solution
{
public:
  string removeDuplicates(string S)
  {
    vector<char> seen;
    seen.push_back(S[0]);

    for (int i = 1; i < S.size(); ++i)
    {
      if (!seen.empty() && seen.back() == S[i])
        seen.pop_back();
      else
        seen.push_back(S[i]);
    }

    return string(seen.begin(), seen.end());
  }
};